ML Case Study Inteview (#51)
agude/agude.github.io: _drafts/2021-01-18-data_science_interview_prep_case_study.md

A common interview type for data scientists and machine learning engineers is
the machine learning case study. In it, the interviewer will ask a question
about how the candidate would build a certain model. These questions can be
challenging for new data scientists because the interview is open-ended and
new data scientists often lack practical experience building and shipping
models at a company.

I have a lot of practice with these types of interviews, both from my time at
[Insight][should_i_go], from my many experiences [interviewing for
jobs][interviewing], and from designing and implementing Intuit's data science
interview. Like my last article where I [put together an example data
manipulation interview practice problem][last_post], this time I will walk
through a practice case study and how I would work through it.

[should_i_go]: {% post_url 2018-08-21-should_i_go_to_insight %}
[interviewing]: {% post_url 2020-09-21-interviewing_for_data_science_positions_in_2020 %}
[last_post]: {% post_url 2020-12-14-data_science_interview_prep_data_manipulation %}

## My Approach

Case study interviews are just conversations. This can make them tougher than

> they need to be because they lack the obvious structure of a coding interview

Author: The latter.

are unlikely to be spam bots, so I can use them. But verified users are
certainly very different from "normal" good users because they have far more
followers.

It is a safe bet that there are far more good users than spam bots, so
randomly selecting accounts can be used to build a negative label set.

### Features

To build features, it helps to think about what sort of behavior a spam bot
might exhibit, and then try to codify that behavior into features. For
example:

- **Bots can't write truly unique messages**; they must use a template or
  language generator. This should lead to similar messages, so looking at how
  repetitive an account's Tweets are is a good feature.
- **Bots are used because they scale.** They can run all the time and send
  messages to hundreds or thousands (or millions) or users. Number of unique
  Tweet recipients and number of minutes per day with a Tweet sent are likely
  good features.
- **Bots have a controller.** Someone is benefiting from the spam, and they
  have to control their bots. Features around logins might help here like
  number of accounts seen from this IP address or device, similarity of login
  time, etc.

### Model Selection

> I try to start at the simplest model that will work when starting a new

Reviewer: ```suggestion
I try to start with the simplest model that will work when starting a new
```

project. Since this is a supervised classification problem and I have written
some simple features, logistic regression or a forest are good candidates. I
would likely go with a forest because they tend to "just work" and are a
little less sensitive to feature processing.[^processing]

[^processing]:
    If you use [regularization] with logistic regression (and you should) you
    need to scale your features. Random forests do not require this.

[regularization]: https://en.wikipedia.org/wiki/Regularization_(mathematics)

Deep learning is not something I would use here. It's great for image, video,
audio, or NLP, but for a problem where you have a set of labels and a set of
features that you believe to be predictive it is generally overkill.

One thing to consider when training is that the dataset is probably going to
be wildly imbalanced. I would start by down-sampling (since we likely have
millions of events), but would be ready to discuss other methods and trade
offs.

### Validation

> Validation is not too hard at this point. We focus on the offline metric we

Reviewer: ```suggestion
Validation is not too difficult at this point. We focus on the offline metric we
```

decided on above: precision. We don't have to worry much about leaking data
between our holdout sets if we split at the account level, although if we
include bots from the same [botnet][botnet] into our different sets there will
be a little data leakage. I would start with a simple validation/training/test
split with fixed fractions of the dataset.

[botnet]: https://en.wikipedia.org/wiki/Botnet

### Deployment

Since we want to classify an entire account and not a specific tweet, we don't
need to run the model in real-time when Tweets are posted. Instead we can run
batches and can decide on the time between runs by looking at something like
the characteristic time a spam bot takes to send out Tweets. We can add
rate limiting to Tweet sending as well to slow the spam bots and give us more
time to decide without impacting normal users.

For deployment, I would start in **shadow mode**, which I [discussed in detail
in another post][shadow_mode]. This would allow us to see how the model
performs on real data without the risk of blocking good accounts. I would

> track it's performance using our online metrics: spam fraction and ops time

Reviewer: ```suggestion
track its performance using our online metrics: spam fraction and ops time
```

> saved. I can compute these metrics both assuming the model does and does not
> block the flagged accounts and then compare them. If the comparison is

Reviewer: ```suggestion
saved. I would compute these metrics twice, once using the assumption that the model blocks flagged accounts, and once assuming that it does not block flagged accounts, and then compare the two outcomes. If the comparison is
```
